cloud/testserver/stacks.go

```diff
@@ -49,232 +76,234 @@ func (handler *stackHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
-
-		// send a ping every 1s
-		for {
-			handler.mu.RLock()
-			logs := handler.logs[orguuid][stackid][deploymentUUID]
-			newLogs := logs[line:]
-			handler.mu.RUnlock()
-
-			for _, l := range newLogs {
-				fmt.Fprintf(w, "%d [%s] %s %s\n", l.Line, l.Channel, l.Timestamp, l.Message)
-				w.(http.Flusher).Flush()
-				line++
-			}
-			if len(newLogs) == 0 {
-				fmt.Fprintf(w, ".\n")
-				w.(http.Flusher).Flush()
-			}
-			time.Sleep(1 * time.Second)
+		if len(logs) == 0 {
+			fmt.Fprintf(w, ".\n")
+			w.(http.Flusher).Flush()
 		}
+		time.Sleep(1 * time.Second)
 	}
+}
 
-	// fake endpoint
-	// PUT /v1/stacks/{orguuid}/{stackid}
-	if r.Method == "PUT" {
-		stackIDStr := params.ByName("stackid")
-		if stackIDStr == "" {
-			w.WriteHeader(http.StatusBadRequest)
-			return
-		}
-		stackid, err := strconv.Atoi(stackIDStr)
-		if err != nil {
-			w.WriteHeader(http.StatusBadRequest)
-			writeErr(w, err)
-		}
-		bodyData, err := io.ReadAll(r.Body)
-		if err != nil {
-			w.WriteHeader(http.StatusInternalServerError)
-			return
-		}
-
-		justClose(r.Body)
+// PostDeploymentLogs is the POST /deployments/.../logs handler.
+func PostDeploymentLogs(store *cloudstore.Data, w http.ResponseWriter, r *http.Request, p httprouter.Params) {
+	stackIDStr := p.ByName("stackid")
+	if stackIDStr == "" {
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+	stackid, err := strconv.Atoi(stackIDStr)
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		writeErr(w, err)
+		return
+	}
+	orguuid := cloud.UUID(p.ByName("orguuid"))
+	org, found := store.GetOrg(orguuid)
+	if !found {
+		w.WriteHeader(http.StatusNotFound)
+		writeString(w, "organization not found")
+		return
+	}
 
-		var st cloud.StackResponse
-		err = json.Unmarshal(bodyData, &st)
-		if err != nil {
-			w.WriteHeader(http.StatusBadRequest)
-			writeErr(w, err)
-			return
-		}
+	stacks := org.Stacks
+	if stackid < 0 || stackid >= len(stacks) {
+		w.WriteHeader(http.StatusNotFound)
+		writeErr(w, errors.E("stack not found"))
+		return
+	}
+	stack := stacks[stackid]
+	deploymentUUID := cloud.UUID(p.ByName("deployment_uuid"))
 
-		if stackid != st.ID {
-			w.WriteHeader(http.StatusBadRequest)
-			return
-		}
+	bodyData, err := io.ReadAll(r.Body)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
 
-		handler.mu.Lock()
-		defer handler.mu.Unlock()
+	justClose(r.Body)
 
-		if _, ok := handler.stacks[orguuid]; !ok {
-			handler.stacks[orguuid] = make(map[int]cloud.StackResponse)
-		}
-		if _, ok := handler.statuses[orguuid]; !ok {
-			handler.statuses[orguuid] = make(map[int]stack.Status)
-		}
+	var logs cloud.DeploymentLogs
+	err = json.Unmarshal(bodyData, &logs)
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
 
-		handler.stacks[orguuid][stackid] = st
-		handler.statuses[orguuid][stackid] = st.Status
-		w.WriteHeader(http.StatusNoContent)
+	err = store.InsertDeploymentLogs(orguuid, stack.Stack.MetaID, deploymentUUID, logs)
+	if err != nil {
```

Comment:
```suggestion
	if err := store.InsertDeploymentLogs(orguuid, stack.Stack.MetaID, deploymentUUID, logs); err != nil {
```

Reply:
sometimes inline error handling like that makes the code hard to read, when the funcall takes lots of arguments, and it trades readability for a small correctness benefit.
But I will see how it looks this way, thanks!